torchgeo/datasets/eurocrops.py

```diff
@@ -0,0 +1,266 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""EuroCrops dataset."""
+
+import csv
+import os
+from collections.abc import Iterable
+from typing import Any, Callable, Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.figure import Figure
+from rasterio.crs import CRS
+
+from .geo import VectorDataset
+from .utils import (
+    DatasetNotFoundError,
+    check_integrity,
+    download_and_extract_archive,
+    download_url,
+)
+
+
+class EuroCrops(VectorDataset):
+    """EuroCrops Dataset (Version 9).
+
+    The `EuroCrops <https://www.eurocrops.tum.de/index.html>`__ dataset combines "all
+    publicly available self-declared crop reporting datasets from countries of the
+    European Union" into a unified format. The dataset is released under CC BY 4.0 Deed.
+
+    The dataset consists of shapefiles containing a total of 22M polygons. Each polygon
+    is tagged with a "EC_hcat_n" attribute indicating the harmonized crop name grown
+    within the polygon in the year associated with the shapefile.
+
+    If you use this dataset in your research, please follow the citation guidelines at
+    https://github.com/maja601/EuroCrops#reference.
+    """
+
+    base_url = "https://zenodo.org/records/8229128/files/"
+
+    hcat_fname = "HCAT2.csv"
+    hcat_md5 = "b323e8de3d8d507bd0550968925b6906"
+    # Name of the column containing HCAT code in CSV file.
+    hcat_code_column = "HCAT2_code"
+
+    label_name = "EC_hcat_c"
+
+    filename_glob = "*_EC*.shp"
+
+    # Override variables to automatically extract timestamp.
+    filename_regex = r"""
+        ^(?P<prefix>[a-zA-Z]+(?:_[a-zA-Z]+)?)
```

Comment:
Do the prefix and suffix have any meaning? Country code or something? Is there a webpage or file that documents this syntax?

Reply:
I would personally be fine with:
```suggestion
        ^(?P<prefix>[A-Z_]+)
```
if that covers everything. I don't expect to find 4-digit sequences in the filename other than the year.

Reply:
> Do the prefix and suffix have any meaning? Country code or something? Is there a webpage or file that documents this syntax?

The EC21 stands for EuroCrops 2021 but some files just have EC instead.

It looks like the first part is either country or country + region. I updated the regex accordingly.

I didn't find any documentation about the filename format. But it should be "country{_region,}_YYYY_EC{21,}.shp".